contracts/CoordinatedTaker.sol

```diff
@@ -0,0 +1,115 @@
+
+/// @title CoordinatedTaker Contract
+/// @author imToken Labs
+contract CoordinatedTaker is ICoordinatedTaker, Ownable, TokenCollector, EIP712 {
+    using Asset for address;
+    using SafeERC20 for IERC20;
+
+    IWETH public immutable weth;
+    ILimitOrderSwap public immutable limitOrderSwap;
+    address public coordinator;
+
+    mapping(bytes32 => bool) public allowFillUsed;
+
+    constructor(
+        address _owner,
+        address _uniswapPermit2,
+        address _allowanceTarget,
+        IWETH _weth,
+        address _coordinator,
+        ILimitOrderSwap _limitOrderSwap
+    ) Ownable(_owner) TokenCollector(_uniswapPermit2, _allowanceTarget) {
+        weth = _weth;
+        coordinator = _coordinator;
+        limitOrderSwap = _limitOrderSwap;
+    }
+
+    receive() external payable {}
+
+    function setCoordinator(address _newCoordinator) external onlyOwner {
+        if (_newCoordinator == address(0)) revert ZeroAddress();
+        coordinator = _newCoordinator;
+
+        emit SetCoordinator(_newCoordinator);
+    }
+
+    function approveTokens(address[] calldata tokens, address[] calldata spenders) external onlyOwner {
+        for (uint256 i = 0; i < tokens.length; ++i) {
+            for (uint256 j = 0; j < spenders.length; ++j) {
+                IERC20(tokens[i]).safeApprove(spenders[j], Constant.MAX_UINT);
+            }
+        }
+    }
+
+    function withdrawTokens(address[] calldata tokens, address recipient) external onlyOwner {
+        for (uint256 i = 0; i < tokens.length; ++i) {
+            uint256 selfBalance = Asset.getBalance(tokens[i], address(this));
+            if (selfBalance > 0) {
+                Asset.transferTo(tokens[i], payable(recipient), selfBalance);
+            }
+        }
+    }
+
+    function submitLimitOrderFill(
+        LimitOrder calldata order,
+        bytes calldata makerSignature,
+        uint256 takerTokenAmount,
+        uint256 makerTokenAmount,
+        bytes calldata extraAction,
+        bytes calldata userTokenPermit,
+        CoordinatorParams calldata crdParams
+    ) external payable override {
+        // validate fill permission
+        {
+            bytes32 orderHash = getLimitOrderHash(order);
+
+            if (crdParams.expiry < uint64(block.timestamp)) revert ExpiredPermission();
+
+            bytes32 allowFillHash = getEIP712Hash(
+                getAllowFillHash(
+                    AllowFill({ orderHash: orderHash, taker: msg.sender, fillAmount: makerTokenAmount, salt: crdParams.salt, expiry: crdParams.expiry })
+                )
+            );
+            if (!SignatureValidator.isValidSignature(coordinator, allowFillHash, crdParams.sig)) revert InvalidSignature();
+
+            if (allowFillUsed[allowFillHash]) revert ReusedPermission();
+            allowFillUsed[allowFillHash] = true;
+        }
+
+        // collect taker token from user (forward to LO contract without check if isETH)
+        if (!order.takerToken.isETH()) {
```

Comment:
The condition of if is `not ETH` so the other side is `ETH` case. The comment explain two side behavior of this if block.

Reply:
I understand that.👍 Though what I meant is that we are indeed performing `check if isETH`, this is literally what the if condition checks: `if (!order.takerToken.isETH())`. It checks `.isETH()` regardless of result being `ETH` or `not ETH`.

To avoid the subtle confusion, I would suggest to reword it like: `(forward to LO contract without validation if taker token is ETH)`